### DISCUSSION TITLE
Plan for existing PSDSC 1.1 / 2.0 / 3.0 users to migrate to DSC 3.0

### DISCUSSION THREAD

**Discussion by Borgquite** [PowerShell/DSC#529] [General]
### Summary of the new feature / enhancement

So @anmenaga suggested that I raise the questions here that I was asking on the DSC community call last night - here goes!

Currently I have a number of nodes managed with PSDSC 1.1 using Azure Automation State Configuration - all on-premises Windows Server nodes. The question was really around:
- What benefits should I expect from moving PSDSC 1.1 to DSC 3.0 - I understand that newer DSC resources could make this a requirement - but what are the driving forces for someone with this setup, apart from 'staying ahead of the curve'?
- What will moving from PSDSC 1.1 to DSC 3.0 look like, especially when it comes to the deprecation of the LCM? I currently have a lot of config which is in the 'classic' PowerShell compliation / MOF format - including looping through data stored in .PSD1 files to automate standing up many resources with small differences, and composite resources. Will there be migration tools available, how challenging is the migration expected to be in my scenario.

I'm familiar with the desire to move from MOF to JSON / YAML and why that makes sense, and 'get' how the new model will make it easier for tools Chef, Puppet, Ansible to integrate with DSC. But what's the benefits that those of us using PSDSC 1.1 in pull mode with the LCM can expect to reap? Just trying to warn about the sense of mere [fire and motion](https://www.joelonsoftware.com/2002/01/06/fire-and-motion/).

### Proposed technical implementation details (optional)

Making the move DSC 1.1 (including using the LCM) to 3.0 as seamless and pain-free as possible.

**Comment by SteveL-MSFT** [comment]
@Borgquite I converted this to a discussion as it's not really an issue (yet).

My expectation is that eventually we'll see more DSCv3 resources simply because they can be written in the language the developer is already familiar with and ideally, the application exe itself serves as the DSC resource rather than a separate thing.  However, this won't happen for some time and in the interim, people can still use existing PowerShell based resources or author new ones in PowerShell.  The shift from MOF to JSON/YAML means you can use much more tooling that is available and alignment with ARM Template syntax means we'll eventually have Bicep support (for those that want it), but also learning one gives you access to both.

My recommendation is that any configuration you have currently that is working to not try to migrate to 3.0.  New configurations should consider being authored in the new 3.0 model.  There is a separate discussion here about a potential migration PowerShell module for existing configurations.

LCM is currently not in the DSC 3.0 plan, but something similar might be discussed for 3.1+.  However, in many cases Azure Machine Config (once they integrate with DSC3) will serve as your LCM.

It would be helpful for our potential LCM plans if you can be more specific on what features of LCM you rely on and the scenarios on how they are used.

**Reply by Borgquite** [reply]
Thanks @SteveL-MSFT. I guess for me, using Azure Machine Configuration to fulfil the functions of an LCM (i.e. get a config, apply it, return the result to the service) would probably work, given that I hope the cost and features would remain equivalent to Azure State Configuration pull server (I hope they would). I think my main concern would be over how to move existing 1.1 configurations over to 3.0 easily. I looked into the process for Machine Configuration for 2.0 and it looked 'feasable' (if I remember correctly, I would just need to do something about my composite configurations which are server-role based, but the foundation is still PowerShell / MOF based). My concerns were more over what I'm seeing in 3.0 re YAML - what does the workflow / migration process look like when moving from 1.1 to 3.0 Azure Machine Configuration when using long-supported and documented constructs to create a configuration - what tools and support are/will there be for transitioning existing 1.1 PowerShell scripts which use e.g. the following functionality to generate MOFs, to create equivalent DSC 3.0 compatible configurations, in JSON/YAML?

- [Parameters](https://learn.microsoft.com/en-us/powershell/dsc/configurations/add-parameters-to-a-configuration?view=dsc-1.1)
- [Conditional statements and loops](https://learn.microsoft.com/en-us/powershell/dsc/configurations/flow-control-in-configurations?view=dsc-1.1)
- [Composite / nested configurations](https://learn.microsoft.com/en-us/powershell/dsc/configurations/compositeconfigs?view=dsc-1.1)
- [Configuration data](https://learn.microsoft.com/en-us/powershell/dsc/configurations/configdata?view=dsc-1.1)
- [Separate configuration and environment data](https://learn.microsoft.com/en-us/powershell/dsc/configurations/separatingenvdata?view=dsc-1.1)

**Reply by michaeltlombardi** [reply]
Regarding those features:

- Parameters are available in DSCv3 configuration documents via the [`parameters` section][01] and can be accessed in resource instances with the [`parameters()` configuration function][02].
- Conditional statements are possible (may require a custom resource) with the `Microsoft.DSC/Assertion` resource and the [`dependsOn`][03] keyword for a resource instance, as shown in [this example][04].
- There are no built-in configuration functions or group resources for loops at this time, but they could be reasonably implemented as a group resource (which I would prefer compared to extending the engine, personally). Essentially, you would define a resource that looks something like:

  ```yaml
  - type: MyOrganization.Logic/ForEach
    name: Services
    properties:
      resources:
        - type: MyOrganization.MyApp/Service
          name: <name> service
          properties: {}
      data:
        - name: MyServiceFoo
          startupType: automatic
        - name: MyServiceBar
          startupType: automatic
        - name: ForbiddenService
          startupType: disabled
  ```
  
  And that would resolve in the nodegraph and act as a group resource to configure each of the resources defined in the data. This model above is just a quick sketch to give you the idea, any actual implementation would require a lot of thought and care around its model and behavior, but I don't see any reason a group resource couldn't handle this sort of thing. Plus, if the data is defined as a configuration parameter, you could separate the data from the foreach group more readily.
  
- Composite/nested configurations can be mimicked with the `Microsoft.DSC/Include` resource, which resolves another configuration document into the current one.

- Configuration data is a little mixed - on the one hand, you can argue that parameters fulfill the role. On the other hand, because configuration documents are (arguably primarily) static, you can't compile different configuration graphs based on that data without a pre-processing step. Accessing different data per-node, definitely possible with parameters. Different resource instances would be more difficult.

  You could handle this with a pre-processing step, generating however many different configuration documents you need. You could also use assertion groups and `dependsOn` to get something similar, but the functionality would be a little messy.
  
  In my opinion, this feature is better handled by higher order tooling, like Machine Configuration - once I'm trying to represent many nodes with differing requirements, I probably want all the other things that come with higher order tooling. If I couldn't use a higher-order tool, I would instead likely decompose my configurations and use the `Microsoft.Dsc/Include` resource to define the documents that actually configure my machines[^1].
  
[^1]: Actually, I would probably do this anyway from the perspective of testing and maintaining my infrastructure. I'm a bit removed from ops now, but being able to separately version, document, and test chunks of my configuration model was a big deal to me and something that the folks I supported when I was at Puppet struggled with.

[01]: https://learn.microsoft.com/powershell/dsc/reference/schemas/config/parameter?view=dsc-3.0&preserveView=true
[02]: https://learn.microsoft.com/powershell/dsc/reference/schemas/config/functions/parameters?view=dsc-3.0&preserveView=true
[03]: https://learn.microsoft.com/powershell/dsc/reference/schemas/config/resource?view=dsc-3.0&preserveView=true#dependson
[04]: https://learn.microsoft.com/powershell/dsc/reference/schemas/config/functions/resourceid?view=dsc-3.0&preserveView=true#example-2---reference-a-group-resource-as-a-dependency

**Reply by Borgquite** [reply]
OK - it's becoming clear to me that this discussion is mis-titled (now corrected) - it seems like my main issue with migration would be as a result of configuration documents needing to be defined in JSON or YAML - using PowerShell scripting / configuration / data files is no longer be supported.

What is the proposed migration path for someone currently using DSC 1.1, or 2.0, when moving to DSC 3.0? Is it basically 'rewrite every configuration in a third-party tooling, using YAML / JSON'? I'm really not clear what the value proposition for people with existing DSC 1.1 or 2.0 configurations lies here? Again, are you risking just consuming our time with [fire and motion](https://www.joelonsoftware.com/2002/01/06/fire-and-motion/)?

It makes sense to me to try to remove the MOF dependency - however what's wrong with maintaining the ability to create configuration documents using the existing PowerShell script syntax - and instead of generating MOF, the output is valid YAML/JSON instead? If you don't want to do that, take the MOF output of a 1.1 / 2.0 configuration document, and allow us to convert it into DSC 3.0 complaint YAML/JSON - surely that's also possible?

I am struggling to see the benefit of rewriting PowerShell configuration documents written for DSC 1.1 or even 2.0 (as currently [recommended by Microsoft](https://learn.microsoft.com/en-us/azure/virtual-machines/extensions/dsc-overview#available-dsc-versions) right now) in an entirely new language, when the only upside I'm being presented with is 'you'll be able to access new DSC resources written in non-PowerShell language, although none of them have actually been written yet'. How do you plan to get critical mass around this without risking losing / forking the existing community?

**Comment by kilasuit** [comment]
@SteveL-MSFT I know you likely haven't but it worth mentioning that we shouldn't be forgetting about the usecases including disconnected environments and non-azure accessible environments including AWS, GCP or other providers including private clouds especially where we want to sequence the order of events in such a way that allows for the best use of available resources in resource constrained environments (like Developer machines)

Also thanks for the comments in the DSCCommunity call on 19th March around this.

**Comment by miketheitguy** [comment]
DSC's biggest problem for almost everyone has always been the orchestration piece. Particularly around on-premises and disconnected environments. And the extreme anti-Microsoft bias out there to running Azure Policy items.

It seems DSC 3.0 didn't address any of this and focused on the authoring bits more so than the orchestration bits.

Linux folks already have their tools. I don't see any world in which the OSS community is going to adopt DSC simply because resources can now be written in any language they want.

Windows folks are yet again stuck with GPO and all of its limitations because it literally is point and click configuration at scale with automatic remediation built right in to the OS.

**Reply by michaeltlombardi** [reply]
You're right that the higher-order configuration management space is one that PSDSC didn't fully tackle. While we have some open issues around improving support/options for disconnected environments, DSC in its current state isn't a fully comparable tool to ansible or even machine configuration. The problem domain of those higher order tools is several orders of magnitude more complex and complicated than just finding, arranging, and invoking resources. It's an entire field of software with multiple full companies trying to solve the domain in different ways.

Speaking to my own experience[^1], I would say that a huge amount of the value of PSDSC, even v1.1, was around the reusability of resources regardless of your specific toolchain. You can leverage PSDSC resources with Ansible, Chef, Puppet, Salt, and numerous Microsoft tools. PSDSC resources fit relatively well into any toolchain, enabling you to use the higher order tools that fit your context. Maybe that's Azure machine configuration, or maybe it's your existing Ansible setup.

DSC doubles down on that value - DSC resources are easier to author and integrate than PSDSC resources have been, and DSC itself is built from the ground up around integrations, data model contracts, and predictable API surfaces for users and integrating software both.

The goal (again, from my perspective) is to enable higher order tools inside and outside Microsoft to build context-specific solutions for high-level orchestration - what a blue-green deployment model infrastructure team wants and needs will be different from IT pros at a small business or from a team managing vast fleets of not-yet-version-controlled machines.

[^1]: My background was first in operations using PSDSC and the LCM directly, then using PSDSC resources with Chef, and finally as a developer at Puppet working on the Windows resources and PSDSC integration. I worked with configuration management tools a lot over the last decade, but usually involved PSDSC to some degree even when I was working at Puppet.

**Reply by miketheitguy** [reply]
Definitely appreciate the response! And you're right here, there's a lot of investment going into orchestration. The issue here is the open source/free orchestration that's out there for those of us who are admins/engineers that don't want to spend our time writing the orchestration glue. Some do, some don't, and many will be bespoke solutions.

I really think it boils down to 3 layers here:

- The orchestration layer
- The Config-As-Code Admin/Dev/User layer
- The configuration layer (the part that actually understands how to apply the configurations)

What it seems to me is that the DSC team tackled the middle layer and a very tiny portion of the bottom layer (the registry bits), and a compatibility layer for PSDSC resources ; while leaving the rest up to the community. Of course, I know the team is pretty small overall so this isn't really a "blame" situation, more of like my frustration with Microsoft as an entity not tackling the other components of this.

I want the Windows Team to tackle that bottom layer, full force, with open source tooling. Simultaneously, I'd like Microsoft (the entity as a whole) to tackle the top layer, whether that's writing the integration for the other bits (doing the leg work on chef, Ansible, Puppet, etc. to build out DSC yaml files) ; or by building a push-button solution to the problem that Windows Administrators could enjoy out of the box.

Without these components it leaves Administrators such as myself struggling with a patchwork of not-quite-there tooling.

With community-based DSC resources that work somewhat okay, but aren't 100%. But the larger part is the orchestration. Distribution of the resources and the yaml files is a difficult challenge for most administrators who even struggle with understanding git and VSCode.

**Comment by Borgquite** [comment]
> I don't see any world in which the OSS community is going to adopt DSC simply because resources can now be written in any language they want.

I have to say there's a lot of sense here. The strength of DSC historically is that it's Microsoft's Windows-native declarative state configuration tool. I can understand the need to make it easier for third-party orchestration tools to call it (e.g. eliminating the MOF dependency). I assume that any pressure for resources to be written in any languages primarily comes from pressure from development teams from within Microsoft. It's hard to imagine what would induce a vendor outside the Microsoft system (like VMWare) choosing to have DSC as their [configuration as code platform](https://github.com/vmware/dscr-for-vmware) in the future, now that Ansible is available.

**Reply by ryanspletzer** [reply]
The thing about Ansible is that it doesn't work great with Windows as a control node...

https://blog.rolpdog.com/2020/03/why-no-ansible-controller-for-windows.html

I'm back into exploring DSC 3.0 (having used prior DSC in the past) as an alternative to Ansible for the likes of dev machine setups on Windows. I wrote a dev machine setup in Ansible for macOS with a simple bash wrapper to invoke (fully parameterized with an input vars yaml file for flexibility), then wanted to do the same for Windows, but to my dismay found the above out the hard way, understanding that I'd have to kick out to some other means like WSL to connect to the host with WinRM with the password, just to configure a localhost Windows machine... which seemed like a lot of extra effort for something simple.

I've done plenty of dev machine setups just with raw powershell scripts that can take in Yaml as a sort of "ConfigurationData" to loop through things (so that people could customize the Yaml input with varying sets of packages like for Homebrew / Chocolatey / apt / yum / pipx / whatever.

In that veign, it would be amazing to have back the ability to loop through an arbitrary set of ConfigurationData.

I'm working around this right now by literally just generating a DSC Yaml config on-the-fly through a PowerShell object export with powershell-yaml, which accepts a Yaml config file as input -- a little corny since you ask yourself "We'll by this point I could have just done this in the PowerShell script instead of kicking out to DSC, since I'm already in a script," but hey it works, and I'm having fun learning. :)

**Reply by Borgquite** [reply]
> In that veign, it would be amazing to have back the ability to loop through an arbitrary set of ConfigurationData.

Indeed. I feel like the @michaeltlombardi use case (using DSC as a low-level implementation allowing it to be called by other higher-level tools) is one helpful way that DSC can be used, but for me it's never been about that. It's a **Windows native** (and I can't stress that enough) way to do things, without having to rely on third-party tools. The big enterprises are going to use their expensive Chef and Puppet and Salt to do all sorts of wonderful configurations managing thousands of nodes, and they've got the cash and the people to do that. However DSC 1.1 (orchestration, configuration creation and implementation) does most of what is needed for small-medium business deployments. It does feel like Microsoft has focussed efforts on making it easy for the 'big guys' like Chef/Puppet/Salt to reuse DSC in their products - (even though as mentioned they're already part of a huge industry that therefore should already *have* the resources to make DSC integrate with their products, or implement their own native resources, so don't really need Microsoft's help to benefit from the hard work of the volunteers in the DSC community) - at the expense of the small/medium business that could really use a Microsoft-native, limited-functionality IaC stack.

**Reply by michaeltlombardi** [reply]
> It's hard to imagine what would induce a vendor outside the Microsoft system (like VMWare) choosing to have DSC as their [configuration as code platform](https://github.com/vmware/dscr-for-vmware) in the future, now that Ansible is available.

I can speak to this a bit from my time at Puppet - for every higher-order configuration management tool, there are `n` implementations of resources to configure vendor software. Consider nginx, which has implementations in [Puppet][01], [Chef][02], and [Ansible][03]. Right now, to configure any software, _someone_ needs to implement the resources for your preferred toolchain - sometimes the vendor does it, sometimes the CM tool company does it, sometimes the community does it, sometimes _you_ need to do it.

One of the benefits of DSC with the data contracts and manifests is that defining a resource to adhere to that contract immediately makes it available to any toolchain. From a software vendor perspective (like Microsoft), that's _very_ compelling. If I owned, say, SQLServer, my team _already_ has to write and maintain how you configure our software. If we expose that configuration through this standard then not only can all of the Microsoft tools leveraging DSC pick it up and enable my users, but all of my users with a completely different toolchain can configure my software the same way - and it's write-once, use-anywhere.

The moonshot aspiration for this model is that software vendors take ownership of their own configurable resources and make them available directly.

> at the expense of the small/medium business that could really use a Microsoft-native, limited-functionality IaC stack.

I think the first thing to remember is that PSDSC is going to live on as long as Windows PowerShell, which I suspect might be "functionally forever." If using PSDSC resources and the LCM compellingly solves your configuration management problems, great! It's more important to get your machines configured than which toolchain you use to do it. My go-to response before, during, and after my time at Puppet to the question "which CM tool should I be using?" was "the one you can actually adopt, implement, and get your team onboarded to."

The second thing to remember is that DSC is very new and rapidly iterating, but that the _plan_ is to use it to power higher order tools from Microsoft, like WinGet and Mahine Configuration and others. DSC isn't a drop-in replacement for PSDSC and the LCM because it's built to enable a new generation of context-specific higher order tools while ensuring that resources developed for any tool are reusable with every other tool.

[01]: https://forge.puppet.com/modules/puppet/nginx/readme
[02]: https://supermarket.chef.io/cookbooks/nginx
[03]: https://github.com/nginx/ansible-role-nginx-config

**Reply by Borgquite** [reply]
Hey @michaeltlombardi 

> I think the first thing to remember is that PSDSC is going to live on as long as Windows PowerShell, which I suspect might be "functionally forever." If using PSDSC resources and the LCM compellingly solves your configuration management problems, great! It's more important to get your machines configured than which toolchain you use to do it. My go-to response before, during, and after my time at Puppet to the question "which CM tool should I be using?" was "the one you can actually adopt, implement, and get your team onboarded to."

I'd love to believe that the PSDSC 1.1 will always be available, but since Microsoft are already deprecating 1.1-based Azure Automation State Configuration (and with the replacement PSDSC 2.0-based Azure Machine Configuration already [missing a number of essential features](https://learn.microsoft.com/en-us/powershell/dsc/overview?view=dsc-2.0#changes-from-dsc-11) that have yet to be replaced - credentials ([secrets management](https://learn.microsoft.com/en-us/azure/governance/machine-configuration/whats-new/migrating-from-dsc-extension#major-differences)), rebooting ([reboot handling](https://learn.microsoft.com/en-us/azure/governance/machine-configuration/whats-new/psdsc-in-machine-configuration#common-dsc-features-not-available-during-machine-configuration-public-preview)), ConfigurationData (it's possible to [pass limited configuration data using string parameters, but not arrays](https://learn.microsoft.com/en-us/azure/governance/machine-configuration/whats-new/psdsc-in-machine-configuration#parameter-support-through-azure-resource-manager)), composite configurations, etc) it's not hard to imagine a future Microsoft manager deciding that PSDSC 1.1 can be ditched and requiring it to be removed from a future OS. After all, it's going to be waay more disruptive to remove [VBScript](https://techcommunity.microsoft.com/blog/windows-itpro-blog/vbscript-deprecation-timelines-and-next-steps/4148301) than PSDSC 1.1, and that's happening!

It doesn't help that MSDSC 3.0 'shares the name' of PSDSC 1.1, despite containing a fraction of the same functionality - but it's easy to be deceived by the 'higher' version number, and assume that it can do everything DSC 1.1 can do, and perhaps more, although the opposite is true.

There are already [rumours of not-too-distant](https://github.com/microsoft/Microsoft365DSC/issues/5140) deprecation / removal of PSDSC in the near future. This despite the advice that Microsoft still offer even today that '[If you aren't using Azure machine configuration, you should use DSC 1.1](https://learn.microsoft.com/en-us/powershell/dsc/overview?view=dsc-2.0#use-case-for-dsc-20)'. If there's reassurance to be offered from Microsoft that PSDSC 1.1 won't be ditched, unless/until there is some sort of feature parity  provided via MSDSC, that could go a long way. Until that happens, those of us running on-premises, will continue to express concern!

**Reply by michaeltlombardi** [reply]
> it's not hard to imagine a future Microsoft manager deciding that PSDSC 1.1 can be ditched and requiring it to be removed from a future OS. 

PSDSC is part of Windows PowerShell - as long as Windows PowerShell continues to ship in Windows, so will PSDSC. If Windows PowerShell is ever going away, it's not going to be in the next five years. There are no known plans to remove Windows PowerShell.

DSC continues to see active development alongside work with partners to integrate DSC into their higher order tools. I'm not saying you shouldn't express concerns, I'm just trying to address your concerns earnestly and realistically. If PSDSC works for you, and DSC isn't compelling enough to get you to switch, I don't think you should feel any pressure to make that switch. Please do continue to share your thoughts and file issues, because DSC is far from "done" and we want it to evolve to serve various communities.
